lec12-numerical-models.Rmd

```diff
@@ -0,0 +1,902 @@
+*First-order* means that the derivatives are all first derivatives (as opposed to second derivatives, third
+derivatives, etc.). **Ordinary** means that the derivatives are full derivatives (as opposed to partial
+derivatives which are written $\partial N / \partial t$). 
+
+This means that the function we'll be using is called `ode`, which stands for 'ordinary differential equation'. 
+
+```{r, eval=FALSE}
+?ode
+# The possible methods are all different ways of numerically solving DEs.
+# stiff vs. non-stiff : no precise definition, but 'stiff' generally means it's possible to be numerically unstable if the step size is too big. 
+```
+
+The method used by `ode` to numerically solve differential equations is called `lsoda`, which automatically
+tries to use the best method depending on the equation. This is why people like to use functions like `ode`:
+the result is often more accurate than Euler's method and might also be faster to run. 
+
+```{r}
+# Example: deSolve with logistic eqn
+
+# now we need to define our function a bit differently to be in the format that `ode` uses
+logistic_fn <- function(t, state, parameters) {
+  # Calculates dN/dt for the logistic equation
+  
+  # t: time point at which to evaluate derivative (doesn't actually change anything in this example)
+  # state: vector of variables (here it's just N)
+  # parameters: vector of model parameters c(r, K)
+  
+  N <- state 
+  
+  r <- parameters[1] # the first element of the parameters vector is r
+  K <- parameters[2] # the second element of the parameters vector is K
+
+  #rate of change
+  dN <- r * N * (1 - N / K)
+    
+  #return rate of change
+  return(list(c(dN)))
+}
+
+parameters <- c(r = 0.5, K = 50)
+state <- c(N = 10)
+times <- seq(0, 50, by = 0.01) # the timestep dt is chosen by setting the increment with 'by'
+
+#?ode # look at the documentation to learn about the parameters
+result <- ode(y = state, times = times, func = logistic_fn, parms = parameters)
+```
+
+The output of `ode` is a matrix that contains the times we requested and its calculated values of N. 
+
+Notice that by using `ode` we didn't have to write a loop explicitly ourselves. Under the hood, `ode` performs a similar computation to what we did manually. 
+
+```{r}
+head(result, 5) # result has a list of times and values of N at those times
+class(result) # result is a 'matrix': an array of numbers
+```
+
+```{r}
+
+result <- data.frame(result) # convert it to a dataframe so we can use ggplot
+
+ggplot(result) +
+  geom_point(aes(x = time, y = N))
+```
+
+## Extras 
+
+### Euler's method in R
+
+Let's review `for loops` first. A loop is a method to do something over and over again, which is
+perfect for Euler's method --- on paper, we were performing the same calculation over and over 
+with a different value of $N$ each time. 
+
+Here's an example from an earlier lecture:
+
+```{r}
+# For each number in v, print the number.
+v <- c(2, 4, 6)
+for (num in v) {
+    print(num)
+}
+```
```

Comment:
Just FYI, I had them do `for loops` in lecture 9 and assignment 5. Nonetheless I'm sure a refresher will be appreciated.

Reply:
:+1: